### PR TITLE
FPGA Register IO - Read testing

### PR DESCRIPTION
**Pull request by DStavilaNI**
- [x] This contribution adheres to [CONTRIBUTING.md](https://github.com/ni/niveristand-fpga-addon-custom-device/blob/master/CONTRIBUTING.md).

### What does this Pull Request accomplish?

- Added "FPGA Register IO" library to the main FPGA Addon repo.
- Implemented tests for "Read" API of the FPGA Register IO library

### Why should this Pull Request be merged?

- Required for the Extended FXP Support feature.

### What testing has been done?

- Manual testing on RT targets has been done, as well as using VI Tester

### CODE REVIEW

**Comment by bogdanp-ni** [comment]
Why is "Source/FPGA Addon System Definition API.lvproj" being deleted (along with other related files)?

File: Source/Tests/Unit Tests/FPGA Register IO Testing/FPGA Register IO Tests/targets.ini [hunk 1]
```diff
@@ -0,0 +1,5 @@
+[Supported Targets]
```
Comment (bogdanp-ni) [inline Source/Tests/Unit Tests/FPGA Register IO Testing/FPGA Register IO Tests/targets.ini:1]: Please add all available targets, not only PharLap.

**Comment by bogdanp-ni** [comment]
FPGA Register IO Tests.lvclass--Get Conifg File Path.vi.png -> typo ("Config")

**Comment by bogdanp-ni** [comment]
1. Could you please explain why we cannot rely on the existing infrastructure that computes the target addresses, like for other tests?
2. This VI will require updating for all available targets and be more flexible when it comes to specifying the RIO address. A simple reconfiguration of the test machine will cause test failures.

image

**Comment by bogdanp-ni** [comment]
As part of the Assets folder, please add the FPGA project and VIs used for generating the bitfiles. Otherwise, I'm just assuming what the tests are doing.

**Comment by bogdanp-ni** [comment]
Please add unit tests (without hw deployments) for the "FXP Raw to DBL" and "DBL to FXP Raw" conversion VIs. These are algorithms where a bug can easily creep in, so we need them thoroughly tested in isolation.

File: Source/Tests/Unit Tests/FPGA Register IO Testing/FPGA Register IO Tests/targets.ini (hunk 1 above)
Comment (zszilagy) [reply Source/Tests/Unit Tests/FPGA Register IO Testing/FPGA Register IO Tests/targets.ini:1]: These are only unit tests which are require an FPGA target. The Controller for these tests are not important. Running it twice will only add to the overal runtime of the ats with no benefit at all.

Comment (zszilagy) [reply Source/Tests/Unit Tests/FPGA Register IO Testing/FPGA Register IO Tests/targets.ini:1]: I left PharLap as it's the first OS which gets installed on the ATS machine.

**Comment by zszilagy** [comment]
> 1. Could you please explain why we cannot rely on the existing infrastructure that computes the target addresses, like for other tests?
> 2. This VI will require updating for all available targets and be more flexible when it comes to specifying the RIO address. A simple reconfiguration of the test machine will cause test failures.
> 
> image

We are running these VIs on the FPGA target outside the VeriStand context. (Just simply running them from LabVIEW). These tests are similar to the LV FPGA tests the NIR VnV team used to run during release validations.

**Comment by bogdanp-ni** [comment]
> > 1. Could you please explain why we cannot rely on the existing infrastructure that computes the target addresses, like for other tests?
> > 2. This VI will require updating for all available targets and be more flexible when it comes to specifying the RIO address. A simple reconfiguration of the test machine will cause test failures.
> > 
> > image
> 
> We are running these VIs on the FPGA target outside the VeriStand context. (Just simply running them from LabVIEW). These tests are similar to the LV FPGA tests the NIR VnV team used to run during release validations.

Makes sense to run them outside the VS context. How do you plan on managing the target addresses?

**Comment by bogdanp-ni** [comment]
FPGA Register IO.lvlib--Read FXP.vi.png -> I would expect the Default case (when Word Length > 64) for the innermost Case Structure to throw an error, not treat the number as a U8. 

Also, I would expect an error if Word Length = 0 (since it doesn't describe a correctly formed FXP number). Are these error cases covered elsewhere in the hierarchy? If not, we should have code guarding for them here.

**Comment by bogdanp-ni** [comment]
In the Write Register polyVI, please populate the lists for the selectors with short, meaningful strings:
image

**Comment by bogdanp-ni** [comment]
Same as above for Read Register poly VI.

**Comment by bogdanp-ni** [comment]
Write U8 Array -> this VI is about U8 arrays, however one of the controls is referring to Cluster Blobs. If this is specific to clusters, then rename the VI. However, it seems to be that this is a generic VI that can be used outside the cluster use-case, so a rename of the control is best.

image

**Comment by bogdanp-ni** [comment]
Make sure that all public API VIs are following the naming conventions for controls and indicators. I saw a few cases where changes are warranted.

image

**Comment by bogdanp-ni** [comment]
I like the idea of caching using the Data Store. However, I would have expected the session ID to invalidate that Data Store. I'm thinking of a case when the session ID changes (or the user has multiple sessions opened), but the Data Store is persisted (in subsequent runs of the Find Register Offset VI), which can corrupt the cache.

**Comment by bogdanp-ni** [comment]
Parse Cluster Configuration.vi -> same question regarding caching applies here, as well.

**Comment by bogdanp-ni** [comment]
Parse Cluster Configuration.vi -> what happens if the cluster contains another cluster? It seems to me that such a case is not supported, which I find rather limiting. Is there a reason why you haven't added nested cluster support?

**Review by bogdanp-ni — CHANGES_REQUESTED** [review]
There are a few outstanding issues. I haven't looked at the actual tests in detail, I'd like to see the FPGA side first, but did a first pass through the rest of the code.

**Comment by DStavilaNI** [comment]
> 
> 
> Why is "Source/FPGA Addon System Definition API.lvproj" being deleted (along with other related files)?

It is no longer necessary since we have the Scripting API for this CD .... as far as I could tell, the System Definition API was just a basic scripting tool that would allow the user to add the FPGA Addon CD to a System Definition.

**Comment by DStavilaNI** [comment]
> 
> 
> FPGA Register IO Tests.lvclass--Get Conifg File Path.vi.png -> typo ("Config")

Done

**Comment by DStavilaNI** [comment]
> 
> 
> As part of the Assets folder, please add the FPGA project and VIs used for generating the bitfiles. Otherwise, I'm just assuming what the tests are doing. I will be looking at the tests after this comment is resolved.

Done

**Comment by bogdanp-ni** [comment]
> > Why is "Source/FPGA Addon System Definition API.lvproj" being deleted (along with other related files)?
> 
> It is no longer necessary since we have the Scripting API for this CD .... as far as I could tell, the System Definition API was just a basic scripting tool that would allow the user to add the FPGA Addon CD to a System Definition.

I agree that it overlaps in functionality with the Scripting API, but it is still part of the CD's public API. Deleting it would break users that have been using it in the past and are now upgrading to a newer version of the CD. At least until we better understand how many such users there are, we must not delete this.

**Comment by DStavilaNI** [comment]
> 
> 
> > > Why is "Source/FPGA Addon System Definition API.lvproj" being deleted (along with other related files)?
> > 
> > 
> > It is no longer necessary since we have the Scripting API for this CD .... as far as I could tell, the System Definition API was just a basic scripting tool that would allow the user to add the FPGA Addon CD to a System Definition.
> 
> I agree that it overlaps in functionality with the Scripting API, but it is still part of the CD's public API. Deleting it would break users that have been using it in the past and are now upgrading to a newer version of the CD. At least until we better understand how many such users there are, we must not delete this.

I would still maintain that it is unnecessary. It is only a single VI ("Create FPGA Addon.vi") and it is not even part of a palette or anything in the LabVIEW IDE. As far as know, it is simply placed in the VS CD directory. The only situation where this could break anything would be if somebody is using that VI directly from that location in his LV Code, however, even then, the Scripting API installer would install an equivalent "Create FPGA Addon.vi" within the LV IDE - this also means that in that case, the customer's code would be able to find the new VI and probably just show some warning. 

I think that it is only normal during software development to deprecate stuff - API is always getting obsolete and replaced and the users notified.

**Comment by DStavilaNI** [comment]
> 
> 
> FPGA Register IO.lvlib--Read FXP.vi.png -> I would expect the Default case (when Word Length > 64) for the innermost Case Structure to throw an error, not treat the number as a U8.
> 
> Also, I would expect an error if Word Length = 0 (since it doesn't describe a correctly formed FXP number). Are these error cases covered elsewhere in the hierarchy? If not, we should have code guarding for them here.

Done

**Comment by DStavilaNI** [comment]
> 
> 
> Please add unit tests (without hw deployments) for the "FXP Raw to DBL" and "DBL to FXP Raw" conversion VIs. These are algorithms where a bug can easily creep in, so we need them thoroughly tested in isolation.

Done

**Comment by DStavilaNI** [comment]
> 
> 
> In the Write Register polyVI, please populate the lists for the selectors with short, meaningful strings:
> image

I do not think that would be necessary. The names of poly instances are already short and readable enough when displayed on the poly selector ... I do not see how this could be improved, it would just mean unnecessary/duplicate work. 

image

Another thing - this is not intended to be a public API and it's used internally within the CD.

**Comment by DStavilaNI** [comment]
> 
> 
> Same as above for Read Register poly VI.

Same observation as above, for the other poly VI

**Comment by bogdanp-ni** [comment]
> > In the Write Register polyVI, please populate the lists for the selectors with short, meaningful strings:
> > image
> 
> I do not think that would be necessary. The names of poly instances are already short and readable enough when displayed on the poly selector ... I do not see how this could be improved, it would just mean unnecessary/duplicate work.
> 
> image


I agree, the polyVI instances look good. No changes needed. I was under the impression that without specifying the Menu and Selector names, the ".vi" extension will be visible, which I wanted to avoid.

> Another thing - this is not intended to be a public API and it's used internally within the CD.

As you previously mentioned in our discussions, this library might turn into a public API. I am reviewing it with that thought in mind, to have it ready when you make the transition.

**Comment by DStavilaNI** [comment]
> 
> 
> Write U8 Array -> this VI is about U8 arrays, however one of the controls is referring to Cluster Blobs. If this is specific to clusters, then rename the VI. However, it seems to be that this is a generic VI that can be used outside the cluster use-case, so a rename of the control is best.
> 
> image

done

> 
> 
> Make sure that all public API VIs are following the naming conventions for controls and indicators. I saw a few cases where changes are warranted.
> 
> image

done

**Comment by DStavilaNI** [comment]
> 
> 
> I like the idea of caching using the Data Store. However, I would have expected the session ID to invalidate that Data Store. I'm thinking of a case when the session ID changes (or the user has multiple sessions opened), but the Data Store is persisted (in subsequent runs of the Find Register Offset VI), which can corrupt the cache.

Technically you are right., This could happen if the user would change the bitfile for example. However I do not think that this is an issue in the FPGA Addon case since the user cannot dynamically, at runtime change the bitfile or the FPGA target - these kind of operations would both require redeployment. In turn, the redeployment would also reset the content of the data store - that would be essentially a new execution.

 At the same time, if the library would be to become stand-alone tool at some point in the future, then yes, this should be addressed.

**Comment by DStavilaNI** [comment]
> 
> 
> Parse Cluster Configuration.vi -> same question regarding caching applies here, as well.

Answered above.

**Comment by DStavilaNI** [comment]
> 
> 
> Parse Cluster Configuration.vi -> what happens if the cluster contains another cluster? It seems to me that such a case is not supported, which I find rather limiting. Is there a reason why you haven't added nested cluster support?

I think that we did discuss this at some point during the stand-ups. Currently, the library does not support nested clusters (or arrays) - this scenario is handled and the API will throw the error 6004 -> Unsupported cluster field datatype.

**Comment by bogdanp-ni** [comment]
> > I like the idea of caching using the Data Store. However, I would have expected the session ID to invalidate that Data Store. I'm thinking of a case when the session ID changes (or the user has multiple sessions opened), but the Data Store is persisted (in subsequent runs of the Find Register Offset VI), which can corrupt the cache.
> 
> Technically you are right., This could happen if the user would change the bitfile for example. However I do not think that this is an issue in the FPGA Addon case since the user cannot dynamically, at runtime change the bitfile or the FPGA target - these kind of operations would both require redeployment. In turn, the redeployment would also reset the content of the data store - that would be essentially a new execution.
> 
> At the same time, if the library would be to become stand-alone tool at some point in the future, then yes, this should be addressed.

Understood. Please file a Tech Debt item for covering the "register offset cache" and "cluster cache", which we need to address before releasing this as a standalone library.

**Comment by bogdanp-ni** [comment]
> > Parse Cluster Configuration.vi -> what happens if the cluster contains another cluster? It seems to me that such a case is not supported, which I find rather limiting. Is there a reason why you haven't added nested cluster support?
> 
> I think that we did discuss this at some point during the stand-ups. Currently, the library does not support nested clusters (or arrays) - this scenario is handled and the API will throw the error 6004 -> Unsupported cluster field datatype.

I admit I don't remember that conversation, and I agree it's not necessary for using this library in FPGA Addon. However, for a separate public release of the library, this is necessary. Please add another Tech Debt covering "nested cluster support" that needs to be done before the standalone release of the library.

**Comment by bogdanp-ni** [comment]
> > > > Why is "Source/FPGA Addon System Definition API.lvproj" being deleted (along with other related files)?
> > > 
> > > 
> > > It is no longer necessary since we have the Scripting API for this CD .... as far as I could tell, the System Definition API was just a basic scripting tool that would allow the user to add the FPGA Addon CD to a System Definition.
> > 
> > 
> > I agree that it overlaps in functionality with the Scripting API, but it is still part of the CD's public API. Deleting it would break users that have been using it in the past and are now upgrading to a newer version of the CD. At least until we better understand how many such users there are, we must not delete this.
> 
> I would still maintain that it is unnecessary. It is only a single VI ("Create FPGA Addon.vi") and it is not even part of a palette or anything in the LabVIEW IDE. As far as know, it is simply placed in the VS CD directory. The only situation where this could break anything would be if somebody is using that VI directly from that location in his LV Code, however, even then, the Scripting API installer would install an equivalent "Create FPGA Addon.vi" within the LV IDE - this also means that in that case, the customer's code would be able to find the new VI and probably just show some warning.
> 
> I think that it is only normal during software development to deprecate stuff - API is always getting obsolete and replaced and the users notified.

For reference, after talking with the PO, we agreed to remove the old scripting node (and project), but as part of a future separate PR.

**Comment by DStavilaNI** [comment]
> 
> 
> > > > > 1. Could you please explain why we cannot rely on the existing infrastructure that computes the target addresses, like for other tests?
> > > > > 2. This VI will require updating for all available targets and be more flexible when it comes to specifying the RIO address. A simple reconfiguration of the test machine will cause test failures.
> > > > > 
> > > > > image
> > > > 
> > > > 
> > > > We are running these VIs on the FPGA target outside the VeriStand context. (Just simply running them from LabVIEW). These tests are similar to the LV FPGA tests the NIR VnV team used to run during release validations.
> > > 
> > > 
> > > Makes sense to run them outside the VS context. How do you plan on managing the target addresses?
> > 
> > 
> > Well RIO1 is the address of the FPGA board in the ATS machine. Hardcoding it and updating it in the VI is ok, but we could also include it in the ini file under a specific section (as we have to parse the ini anyway)
> 
> Please avoid hardcoding RIO addresses inside a VI. That's what the INI should be for.

Done

**Comment by DStavilaNI** [comment]
> 
> 
> > > I like the idea of caching using the Data Store. However, I would have expected the session ID to invalidate that Data Store. I'm thinking of a case when the session ID changes (or the user has multiple sessions opened), but the Data Store is persisted (in subsequent runs of the Find Register Offset VI), which can corrupt the cache.
> > 
> > 
> > Technically you are right., This could happen if the user would change the bitfile for example. However I do not think that this is an issue in the FPGA Addon case since the user cannot dynamically, at runtime change the bitfile or the FPGA target - these kind of operations would both require redeployment. In turn, the redeployment would also reset the content of the data store - that would be essentially a new execution.
> > At the same time, if the library would be to become stand-alone tool at some point in the future, then yes, this should be addressed.
> 
> Understood. Please file a Tech Debt item for covering the "register offset cache" and "cluster cache", which we need to address before releasing this as a standalone library.

Done: https://ni.visualstudio.com/DevCentral/_boards/board/t/ASW%20G%20Ecosystem/Work%20Items/?workitem=1451342

**Comment by DStavilaNI** [comment]
> 
> 
> > Please add unit tests (without hw deployments) for the "FXP Raw to DBL" and "DBL to FXP Raw" conversion VIs. These are algorithms where a bug can easily creep in, so we need them thoroughly tested in isolation.
> 
> Were these tests added? I don't see any new tests. I am ok with getting them in in a separate PR, but please make them a priority (I see them as mandatory for closing the User Story).

The tests are available - > 
image

**Review by bogdanp-ni — APPROVED** [review]
This one's good to go.

I will add a bug under this feature for having more coverage on the conversion code (FXP <-> DBL). I will list all testcases that I want covered. In the interest of time, we'll be adding those testcases in another PR.

**Pull request merged**